Improve Template Diagnostics

zarend opened the pull request:
This PR has two commits, the first is preliminary work to fix broken compliance tests that blocked the main work, which is improving our template diagnostics in language service.

### Improve Template Diagnostics

Durring analysis we find template parse errors. This commit changes
where the type checking context stores the parse errors. Previously, we
stored them on the AnalysisOutput this commit changes the errors to be
stored on the TemplateData (which is a property on the shim). That way,
the template parse errors can be grouped by template.

Previously, if a template had a parse error, we poisoned the module and
would not procede to find typecheck errors. This change does not poison
modules whose template have typecheck errors, so that ngtsc can emit
typecheck errors for templates with parse errors.

Additionally, all template diagnostics are produced in the same place.
This allows requesting just the template template diagnostics or just
other types of errors.

### Preliminary Work

Refactors the i18n error tests to be unit tests in ngtsc_spec.ts. There
is two reasons for doing this.

First is that the tests in compliace_old expected an expection to be be
thrown but did not fail the test if no exception was thrown. That means
that this test could miss catching a bug. It is also a big hacky to call
compile directly and expect an exception to be thrown for diagnostics.

Also, this can easily be unit tested and an end-to-end test is not
necessary since we are not making use of the goldfiles for these tests.

It is easier to maintain and less hacky to validate that we get helpful
error messages when nesting i18n sections by calling getDiagnostics
directly.

zarend committed:
refactor(compiler-cli): move template parse errors to TemplateData

Durring analysis we find template parse errors. This commit changes
where the type checking context stores the parse errors. Previously, we
stored them on the AnalysisOutput this commit changes the errors to be
stored on the TemplateData (which is a property on the shim). That way,
the template parse errors can be grouped by template.

Previously, if a template had a parse error, we poisoned the module and
would not procede to find typecheck errors. This change does not poison
modules whose template have typecheck errors, so that ngtsc can emit
typecheck errors for templates with parse errors.

Additionally, all template diagnostics are produced in the same place.
This allows requesting just the template template diagnostics or just
other types of errors.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -135,6 +138,7 @@ export interface TypeCheckingHost {
   /**
    * Check if the given component has had its template overridden, and retrieve the new template
    * nodes if so.
+   * TODO(Ayaz): store parse errors from overridden template
```
Just a couple nits:
- I'm wondering if this is the right place for the `TODO` (since it's just an interface). I saw a similar comment below near the actual implementation, should that be sufficient?
- I'd propose to put Ayaz's GitHub handle so that it's easier to find all TODOs later, i.e. something like `TODO(ayazhafiz): ...`.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -205,12 +209,37 @@ export class TypeCheckContextImpl implements TypeCheckContext {
+
+    if (parseErrors !== null) {
+      // TODO(Ayaz): get parse errors from template override
+      templateDiagnostics.push(...parseErrors.map(error => {
```
Given the current size of the `addTemplate` function, I'd propose to move the logic that converts parse errors into template diagnostics into a new private function, so this call would look like:
```
  templateDiagnostics.push(...this.convertParseErrorsToDiagnostic(parseErrors, ref.node));
```

ayazhafiz commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -232,6 +232,10 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
       diagnostics.push(...typeCheckProgram.getSemanticDiagnostics(shimSf).map(
           diag => convertDiagnostic(diag, fileRecord.sourceManager)));
       diagnostics.push(...shimRecord.genesisDiagnostics);
+
+      for (let templateData of shimRecord.templates.values()) {
+        diagnostics.push(...templateData.parseErrors);
+      }
```
Do we want to retrieve typecheck diagnostics for templates that have parse errors? I imagine there may be some invalid/non-sensical typeck diagnostics that can be produced if we pass an invalid (even if well-formed) template to the typechecker. For example, once #39523 lands, a template like

```
<div ()="something"></div>
```

would produce a parse error about an empty event, but given that we still form an AST with an empty event binding name, would the typechecker also give a diagnostic about no event found for the div element?

@alxhub and @atscott probably have more insight here; I am fine lumping both kinds of diagnostic for now and figuring out if/where this leads to poor diagnostic reporting later on, but just something to think about.

zarend replied:
Thanks for review @ayazhafiz . Yes, there might be some confusing errors caused by an upstream template parse error. It is also valuable to show some errors on a template if it has parse errors, and we that that makes this worth it.

For example, here I make parse error by doing an illegal assignment, then I continue to write html that parses okay, but has typechecking errors. It's valuable to show developers errors in the rest of the template, even if another part of the template has a parse error.

```
<input ngModel="foo = $event"/> <!-- Parse error, assignments not allowed -->
<app-cat [name]="bar"></app-cat> <!-- Error, type number is not assigning to type string-->
```

@alxhub , care to weigh in :)

ayazhafiz replied:
I agree it would be good to have both parse and typeck errors at the same time. I mean to say that a parse error could cause us to parse an AST that would also then have a typecheck error over the same span as a parse error, and maybe we don't care about the typecheck error in this case. But I don't think this is something worth trying to "fix" at this point.

ayazhafiz commented on packages/language-service/ivy/test/diagnostic_spec.ts:
```diff
@@ -97,12 +95,84 @@ describe('getSemanticDiagnostics', () => {
+        LanguageServiceTestEnvironment.setup([moduleFile, appFile, templateFile1, templateFile2]);
+
+    const diags = env.ngLS.getSemanticDiagnostics(absoluteFrom('/app.ts'));
+    expect(diags.length).toEqual(2);
+
+    expect(diags.map(x => x.messageText).sort()).toEqual([
+      'Parser Error: Bindings cannot contain assignments at column 8 in [{{nope = false}}] in /app1.html@0:0',
+      'Parser Error: Bindings cannot contain assignments at column 8 in [{{nope = true}}] in /app2.html@0:0'
+    ]);
```
nit: no need to assert the length since the array is matched for equality

ayazhafiz commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -135,6 +138,7 @@ export interface TypeCheckingHost {
   /**
    * Check if the given component has had its template overridden, and retrieve the new template
    * nodes if so.
+   * TODO(Ayaz): store parse errors from overridden template
```
I think you should assign this TODO to yourself just so there is less "handing off" and this work can land faster 🙂

zarend replied:
fair enough, Andrew S. and I will take this on :)

zarend committed:
test(compiler-cli): fix i18n error tests

Refactors the i18n error tests to be unit tests in ngtsc_spec.ts. There
is two reasons for doing this.

First is that the tests in compliace_old expected an expection to be be
thrown but did not fail the test if no exception was thrown. That means
that this test could miss catching a bug. It is also a big hacky to call
compile directly and expect an exception to be thrown for diagnostics.

Also, this can easily be unit tested and an end-to-end test is not
necessary since we are not making use of the goldfiles for these tests.

It is easier to maintain and less hacky to validate that we get helpful
error messages when nesting i18n sections by calling getDiagnostics
directly.

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -205,12 +211,23 @@ export class TypeCheckContextImpl implements TypeCheckContext {
+
+    if (parseErrors !== null) {
+      templateDiagnostics.push(
+          ...(this.getTemplateDiagnostics(parseErrors, templateId, sourceMapping)));
```
nit: this extra parens isn't necessary.

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -251,11 +266,8 @@ export class TypeCheckContextImpl implements TypeCheckContext {
-      boundTarget,
-    });
+
+    shimData.templates.set(templateId, {template, boundTarget, parseErrors: templateDiagnostics});
```
nit: add trailing comma to this object so each property gets its own line

atscott reviewed (changes requested):
What happened to the change to prevent compilation of a component if there are parse errors? and handling the potential for `null` resolution data?

zarend commented:
> What happened to the change to prevent compilation of a component if there are parse errors? and handling the potential for `null` resolution data?

### How do we avoid compiling components with parse errors?

We don't want to do that because we could cause confusing errors from things being in an invalid state. These could look like red herrings when the source of the problem is the parse error. (E.G. if you try to compile a template with nested `i18n` sections, you get `i18n context not available`. Instead we provide a parse error saying `nesting i18n sections is not allowed. Please remove the nesting).

![image](https://user-images.githubusercontent.com/7720245/101948724-64b39500-3ba7-11eb-8890-e8cf96750129.png)
The safeguard from compiling a bad template is handed in packages/compiler-cli/src/perform_compile.ts. If we have any diagnostics, then we refuse the emit the program.

We know there's a diagnostic for template parse errors because we put them there in the NgCompiler. Emit is the phase we don't want to do because it will compile the template.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -232,6 +232,10 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
           diag => convertDiagnostic(diag, fileRecord.sourceManager)));
       diagnostics.push(...shimRecord.genesisDiagnostics);
+
+      for (let templateData of shimRecord.templates.values()) {
```
```suggestion
      for (const templateData of shimRecord.templates.values()) {
```

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -7587,15 +7590,105 @@ export const Foo = Foo__PRE_R3__;
+        });
+      });
+
+      describe('i18n errors', () => {
```
Remove from (new) compliance tests

zarend replied:
✔️ they've been removed in commit `3d7344172f269bf2560dd8ccafeaae8819d441c1`

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -205,12 +211,23 @@ export class TypeCheckContextImpl implements TypeCheckContext {
+
+    if (parseErrors !== null) {
+      templateDiagnostics.push(
+          ...(this.getTemplateDiagnostics(parseErrors, templateId, sourceMapping)));
```
What if the template is overridden? Let's capture and use those parse errors instead here.

zarend replied:
done 👍

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/api/context.ts:
```diff
@@ -40,7 +40,8 @@ export interface TypeCheckContext {
       pipes: Map<string, Reference<ClassDeclaration<ts.ClassDeclaration>>>,
-      schemas: SchemaMetadata[], sourceMapping: TemplateSourceMapping, file: ParseSourceFile): void;
+      schemas: SchemaMetadata[], sourceMapping: TemplateSourceMapping, file: ParseSourceFile,
+      parseErrors: ParseError[]|null): void;
```
It looks like this new params need to be added to JSDoc above.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -64,6 +70,11 @@ export interface TemplateData {
+  /**
+   * Errors found while parsing them template, which have been converted to diagnostics.
+   */
+  parseErrors: TemplateDiagnostic[];
```
Nit: I was thinking if `parseErrors` name here overlap with `parseErrors` in other parts of the code where fields with the same name have `ParseError[]` type. May be it makes sense to call it `errors` or `diagnostics`? Or be a bit more specific and call it `parseErrorDiagnostics`?
(Note: I'm also ok to keep it as is, just sharing some thoughts)

zarend replied:
cool, I named it templateDiagnostics

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -205,24 +216,34 @@ export class TypeCheckContextImpl implements TypeCheckContext {
-      schemas: SchemaMetadata[], sourceMapping: TemplateSourceMapping,
-      file: ParseSourceFile): void {
+      schemas: SchemaMetadata[], sourceMapping: TemplateSourceMapping, file: ParseSourceFile,
+      parseErrors: ParseError[]|null): void {
```
JSDoc for this fn looks outdated (and also missing new arg).

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -294,6 +316,25 @@ export class TypeCheckContextImpl implements TypeCheckContext {
   }
 
+  private getTemplateDiagnostics(
+      parseErrors: ParseError[], templateId: TemplateId, sourceMapping: TemplateSourceMapping) {
```
nit: function is missing its return type.

AndrewKushnir replied:
One more note: it looks like all private methods of the `TypeCheckContextImpl` class go after non-private ones. For consistency, I'd propose moving this method down as well (put it after `finalize` method).

AndrewKushnir reviewed (approved):
LGTM (with a couple minor comments) 👍

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/component.ts:
```diff
@@ -371,9 +349,8 @@ export class ComponentDecoratorHandler implements
           template: templateResource,
         },
-        isPoisoned: diagnostics !== undefined && diagnostics.length > 0,
+        isPoisoned: false,
```
We should make sure to early-exit from `compilePartial` and `compileFull` for components that have parse errors, since we're removing the diagnostics from here, which will allow this component to "proceed" down the line with compilation.

zarend commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -7587,15 +7590,118 @@ export const Foo = Foo__PRE_R3__;
+                 .toEqual('<div i18n>Content</div>');
+           });
+
+        it('report a diagnostic and doesn\'t emit ' +
```
@alxhub would you do this as two seperate `it`'s? One it to check the diagnostics, and another to make sure it doesn't emit the component with a bad template?

zarend commented:
> > What happened to the change to prevent compilation of a component if there are parse errors? and handling the potential for `null` resolution data?
> 
> ### How do we avoid compiling components with parse errors?
> We don't want to do that because we could cause confusing errors from things being in an invalid state. These could look like red herrings when the source of the problem is the parse error. (E.G. if you try to compile a template with nested `i18n` sections, you get `i18n context not available`. Instead we provide a parse error saying `nesting i18n sections is not allowed. Please remove the nesting).
> 
> ![image](https://user-images.githubusercontent.com/7720245/101948724-64b39500-3ba7-11eb-8890-e8cf96750129.png)
> The safeguard from compiling a bad template is handed in packages/compiler-cli/src/perform_compile.ts. If we have any diagnostics, then we refuse the emit the program.
> 
> We know there's a diagnostic for template parse errors because we put them there in the NgCompiler. Emit is the phase we don't want to do because it will compile the template.

^ I was partially correct about this. We also need to early return in `compileFull` and `compilePartial` so that we try to compile a bad template and crash internally. Ty to @alxhub for catching this.

atscott commented:
> We also need to early return in compileFull and compilePartial so that if someone trys to call emit directly, we don't crash internally when trying to compile a bad template. 

@zarend - Do you have any examples of scenarios where this would happen? Incremental builds?

alxhub commented:
> > We also need to early return in compileFull and compilePartial so that if someone trys to call emit directly, we don't crash internally when trying to compile a bad template.
> 
> @zarend - Do you have any examples of scenarios where this would happen? Incremental builds?

If a consumer of the `NgtscProgram` APIs attempts to `emit()` even if there are diagnostics.
